python/ray/local_mode_manager.py

```diff
@@ -0,0 +1,53 @@
+import copy
+import random
+import string
+import traceback
+
+import numpy as np
+
+from ray import ObjectID
+from ray.utils import format_error_message
+from ray.exceptions import RayTaskError
+
+
+def random_object_id():
+    return ObjectID(np.random.bytes(20))
+
+
+class LocalModeManager():
+    def __init__(self):
+        self.object_store = {}
```

Comment:
Will this just keep all objects in scope forever leading to memory usage growing linearly with time? It seems that the only way that objects get removed is when `free` is called. But `ray.internal.free` is never really called (it can be called explicitly by power users, but doesn't happen automatically under the hood).

One possibility might be to use some sort of weakref dictionary, which doesn't increase the reference count on the object.

Reply:
In the test, it might be a good idea to actually check the size of this dictionary to make sure it is the expected size.

Reply:
Would it make sense to store the task return value as an extra field inside of the object ID (instead of in a dictionary)? That might address the issue.

Reply:
That would be a nice, simple solution but I don't think it will cover all cases (i.e., shipping ObjectIDs out-of-band). However, it should be straightforward to notify the user when this happens - we will need a wrapper class (e.g., `LocalObjectID`) anyways to store the return value in an extra field, so if the user passes in an `ObjectID` instead of a `LocalObjectID`, we can raise an error.

I would be happy with that solution, especially given that we'll need to reimplement this in the core worker anyways. What do you think?

Reply:
Yes, I think that sounds good. Especially if we have tests in place that do some amount of checking to make sure memory is not leaked.

Reply:
Actually, I'm not sure how I feel about introducing a `LocalObjectID` wrapper class in part because it is sort of an API difference. E.g., if people use `isinstance` in their code it will behave differently. Couldn't we accomplish the same thing by potentially checking for the presence/absence of some field of the `ObjectID` object (this would require modifying the ObjectID cython class though)?

Also, if it is overly involved, we could defer the solution for now since much of this code will change, and this PR is already a substantial improvement over the previous version.

Reply:
I thought about that as well, but actually the behavior of `isinstance(obj, ObjectID)` won't change if it's a subclass and I think a behavior change if people are using `type()` in their code is acceptable for a debugging mode.

I would prefer this over creating an extra unused field in the ObjectID because it confines the changes to local mode. Don't really want to be add a field to the ObjectID and then possibly remove it once we make the core worker changes.

Reply:
Ok, that seems reasonable.